crates/wasmtime/src/values.rs

```diff
@@ -211,51 +232,47 @@ impl From<Option<ExternRef>> for Val {
     }
 }
 
+impl From<Option<Func>> for Val {
+    fn from(val: Option<Func>) -> Val {
+        Val::FuncRef(val)
+    }
+}
+
 impl From<Func> for Val {
     fn from(val: Func) -> Val {
-        Val::FuncRef(val)
+        Val::FuncRef(Some(val))
     }
 }
 
 pub(crate) fn into_checked_anyfunc(
     val: Val,
     store: &Store,
-) -> Result<wasmtime_runtime::VMCallerCheckedAnyfunc> {
+) -> Result<NonNull<wasmtime_runtime::VMCallerCheckedAnyfunc>> {
     if !val.comes_from_same_store(store) {
         bail!("cross-`Store` values are not supported");
     }
     Ok(match val {
-        Val::ExternRef(None) => wasmtime_runtime::VMCallerCheckedAnyfunc {
-            func_ptr: ptr::null(),
-            type_index: wasmtime_runtime::VMSharedSignatureIndex::default(),
-            vmctx: ptr::null_mut(),
-        },
-        Val::FuncRef(f) => {
-            let f = f.wasmtime_function();
-            wasmtime_runtime::VMCallerCheckedAnyfunc {
-                func_ptr: f.address,
-                type_index: f.signature,
-                vmctx: f.vmctx,
-            }
-        }
+        Val::FuncRef(None) => store.null_func_ref(),
+        Val::FuncRef(Some(f)) => f.caller_checked_anyfunc(),
         _ => bail!("val is not funcref"),
     })
 }
 
-pub(crate) fn from_checked_anyfunc(
-    item: wasmtime_runtime::VMCallerCheckedAnyfunc,
+pub(crate) unsafe fn from_checked_anyfunc(
+    anyfunc: NonNull<wasmtime_runtime::VMCallerCheckedAnyfunc>,
     store: &Store,
 ) -> Val {
-    if item.type_index == wasmtime_runtime::VMSharedSignatureIndex::default() {
-        return Val::ExternRef(None);
+    if anyfunc.as_ref().func_ptr.is_null() {
```

Comment:
Another passing though, perhaps `fn is_null(&self) -> bool` could be added to `VMCallerCheckedAnyfunc`